lib/url.ts

```diff
@@ -96,25 +144,62 @@ export class URLBuilder {
   }
 
   /**
-   * Set the query parameter in this URL with the provided queryParameterName to be the provided
-   * queryParameterEncodedValue.
+   * If the provided searchValue is found in this URLBuilder's path, then replace it with the
+   * provided replaceValue.
    */
-  public setQueryParameter(queryParameterName: string, queryParameterEncodedValue: string): void {
-    if (queryParameterName) {
-      this._query[queryParameterName] = queryParameterEncodedValue;
+  public pathSubstitution(searchValue: string, replaceValue: string): void {
+    if (this._path && searchValue) {
+      this._path = replaceAll(this._path, searchValue, replaceValue || "");
     }
   }
 
   /**
    * Set the query in this URL.
    */
-  public setQuery(query: string | URLQuery | undefined): void {
+  public setQuery(query: string | undefined): void {
     if (!query) {
-      this._query = {};
-    } else if (typeof query !== "string") {
-      this._query = query;
+      this._query = undefined;
     } else {
-      this.set(query, URLTokenizerState.QUERY);
+      this._query = URLQuery.parse(query);
+    }
+  }
+
+  /**
+   * Set a query parameter with the provided name and value in this URL's query. If the provided
+   * query parameter value is undefined or empty, then the query parameter will be removed if it
+   * existed.
+   */
+  public setQueryParameter(queryParameterName: string, queryParameterValue: string | undefined): void {
+    if (queryParameterName) {
+      if (!this._query) {
+        this._query = new URLQuery();
+      }
+      this._query.set(queryParameterName, queryParameterValue);
+    }
+  }
+
+  /**
+   * Get the value of the query parameter with the provided query parameter name. If no query
+   * parameter exists with the provided name, then undefined will be returned.
+   */
+  public getQueryParameterValue(queryParameterName: string): string | undefined {
+    return this._query ? this._query.get(queryParameterName) : undefined;
+  }
+
+  /**
+   * Get the query in this URL.
+   */
+  public getQuery(): string | undefined {
+    return this._query ? this._query.toString() : undefined;
+  }
```

Comment:
Copying from [swagger specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#fixed-fields-7)


collectionFormat | string | Determines the format of the array if type array is used. Possible values are:csv - comma separated values foo,bar.ssv - space separated values foo bar.tsv - tab separated values foo\tbar.pipes - pipe separated values foo\|bar.multi - corresponds to multiple parameter instances instead of multiple values for a single instance foo=bar&foo=baz. This is valid only for parameters in "query" or "formData".Default value is csv.
-- | -- | --

Reply:
Since this is swagger specific, I'd like to keep it outside of the URL logic. I think that the URL classes should keep things simple and in that regard should think about query parameters as just being string to string mappings. Any formatting/conversion should happen outside of these URL specific classes.